src/LanguageServer/Protocol/Protocol/DocumentUri.cs

```diff
@@ -0,0 +1,137 @@
+/// see https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uri
+/// </summary>
+/// <remarks>
+/// While .NET has a type represent URIs (System.Uri), we do not want to use this type directly in
+/// serialization and deserialization.  System.Uri does full parsing and validation on the URI upfront, so
+/// any issues in the uri format will cause deserialization to fail and bypass any of our error recovery.
+/// 
+/// Compounding this problem, System.Uri will fail to parse various RFC spec valid URIs.
+/// In order to gracefully handle these issues, we defer the parsing of the URI until someone
+/// actually asks for it (and can handle the failure).
+/// </remarks>
+internal sealed class DocumentUri : IEquatable<DocumentUri>
+{
+    private readonly Lazy<Uri?> _parsedUriLazy;
+
+    public DocumentUri(string uriString)
+    {
+        UriString = uriString;
+        _parsedUriLazy = new(() => ParseUri(uriString));
+    }
+
+    public DocumentUri(Uri parsedUri)
+    {
+        UriString = parsedUri.AbsoluteUri;
+        _parsedUriLazy = new(() => parsedUri);
+    }
+
+    public string UriString { get; }
+
+    /// <summary>
+    /// Gets the parsed System.Uri for the URI string.
+    /// </summary>
+    /// <returns>
+    /// Null if the URI string is not parse-able with System.Uri.
+    /// </returns>
+    /// <remarks>
+    /// Invalid RFC spec URI strings are not parse-able as so will return null here.
+    /// However, System.Uri can also fail to parse certain valid RFC spec URI strings.
+    /// 
+    /// For example, any URI containing a 'sub-delims' character in the host name
+    /// is a valid RFC spec URI, but will fail with System.Uri
+    /// </remarks>
+    public Uri? ParsedUri => _parsedUriLazy.Value;
+
+    private static Uri? ParseUri(string uriString)
+    {
+        try
+        {
+            return new Uri(uriString);
+        }
+        catch (UriFormatException)
+        {
+            // This is not a URI that System.Uri can handle.
+            return null;
+        }
+    }
+
+    public override string ToString() => UriString;
+
+    public override bool Equals([NotNullWhen(true)] object? obj) => obj is DocumentUri other && this.Equals(other);
+
+    public bool Equals(DocumentUri otherUri)
+    {
+        // 99% of the time the equivalent URIs will have equivalent URI strings, as the client is expected to be consistent in how it sends the URIs to the server,
+        // either always encoded or always unencoded.
+        // See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uri
+        if (this.UriString == otherUri.UriString)
+        {
+            return true;
+        }
+
+        // If either of the URIs cannot be parsed, we'll compare the original URI strings.
+        if (otherUri.ParsedUri is null || this.ParsedUri is null)
+        {
+            return this.UriString == otherUri.UriString;
```

Comment:
this must be false, right?  because we did this exact check above, and would have returned 'true' if it succeeded.  so why not return false here?

Reply:
yup, that should definitely return false.  good catch